R/CDPSession.R

```diff
@@ -1,14 +1,47 @@
 #' @include EventEmitter.R
-NULL
-
-#' Create a websocket connexion using the Chrome DevTools Protocol
-#'
-#' @name CDPSession
+#' @include domain.R
+#' @include CDProtocol.R
+#' @include utils.R
 NULL
 
 #' @export
-CDPSession <- R6::R6Class(
-  "CDPSession",
+CDPSession <- function(
+  host = "localhost", port = 9222, secure = FALSE, ws_url = NULL,
+  autoConnect = FALSE, local = FALSE
+) {
+  url <- build_url(host, port, secure)
+  protocol <- CDProtocol$new(url = url, local = local)
+  if(is.null(ws_url)) {
+    ws_url <- chr_get_ws_addr(port = port)
+  }
+  CDPSession <- R6::R6Class(
+    "CDPSession",
+    inherit = CDPConnexion,
+    public = list(
+      initialize = function(ws_url, protocol, autoConnect) {
+        super$initialize(ws_url = ws_url, autoConnect = FALSE)
+        self$.__protocol__ <- protocol
+        lapply(protocol$domains, function(name) self[[name]] <- domain(self, name))
+        self$.__protocol__ <- protocol
+        if(isTRUE(autoConnect)) {
+          self$connect()
+        }
+      },
+      .__protocol__ = NULL
+    )
+  )
+  lapply(protocol$domains, function(domain) CDPSession$set("public", domain, NULL))
+  client <- CDPSession$new(ws_url = ws_url, protocol = protocol, autoConnect = autoConnect)
+  if(isTRUE(autoConnect)) {
+    while(client$readyState() == 0L) {
+      later::run_now()
```

Comment:
I understand you don't like `later::run_now()` here as it already leave in `await` and all use of `later::run_now()` somewhere should in fact use `await`. Hope I understood it well because I agree with that 😄 

> With a promise, the connection process would be client <- await(remote$connect()) 
Seems logical. And this would be wrapped in a `remote_connect()` function ? 😉

Reply:
This is exactly what I think !